File: pa1/closestpair1.py
import math
import sys

def closest_pair(points):
    if len(points) == 1:
        return ((-sys.maxsize, -sys.maxsize), (sys.maxsize, sys.maxsize))

    if len(points) == 2:
        return points[0], points[1]

    if len(points) == 3:
        return brute_force(points)

    points.sort()

    mid = len(points) // 2
    left_points = points[:mid]
    right_points = points[mid + 1:]

    pair1 = closest_pair(left_points)
    pair2 = closest_pair(right_points)
    dist1 = dist(pair1[0], pair1[1])
    dist2 = dist(pair2[0], pair2[1])

    if dist1 < dist2:
        dist = dist1
        which_pair = 1
    else:
        dist = dist2
        which_pair = 2

    line_points = [point for point in points if abs(point[0] - points[mid][0]) < dist]
    line_points.sort(key=lambda x: x[1])

    pair3 = closest_pair_in_line(line_points, dist)
    dist3 = dist(pair3[0], pair3[1])

    if dist3 < dist:
        dist = dist3
        which_pair = 3

    if which_pair == 1:
        return pair1
    if which_pair == 2:
        return pair2
    if which_pair == 3:
        return pair3

def closest_pair_in_line(points, dist):
    pair3 = ((-sys.maxsize, -sys.maxsize), (sys.maxsize, sys.maxsize))
    min_dist = dist

    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            if points[j][1] - points[i][1] < min_dist:
                if dist(points[j], points[i]) < min_dist:
                    pair3 = points[j], points[i]
                    min_dist = dist(points[j], points[i])

    return pair3

def dist(a, b):
    dist_x = a[0] - b[0]
    dist_y = a[1] - b[1]
    return math.sqrt(dist_x * dist_x + dist_y * dist_y)

def brute_force(points):
    pair3 = ((-sys.maxsize, -sys.maxsize), (sys.maxsize, sys.maxsize))
    min_dist = dist(pair3[0], pair3[1])

    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            if dist(points[j], points[i]) < min_dist:
                pair3 = points[j], points[i]
                min_dist = dist(points[j], points[i])

    return pair3


while True:
    n = int(input())
    if n == 0:
        exit()

    points = []
    for _ in range(n):
        x, y = map(float, input().split())
        points.append((x, y))

    res = closest_pair(points)
    print("{:.4f} {:.4f} {:.4f} {:.4f}".format(res[0][0], res[0][1], res[1][0], res[1][1]))

File: pa1/closestpair1.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <climits>

using namespace std;

struct Point {
    double x, y;
};

pair<Point, Point> closestPair(vector<Point>& points);
double euclideanDistance(const Point& a, const Point& b);
pair<Point, Point> closestPairInLine(vector<Point>& points, double distance);
bool sortByY(const Point& a, const Point& b);
pair<Point, Point> bruteForce(vector<Point>& points);

void readPoints(vector<Point>& points) {
    int n;
    cin >> n;
    points.resize(n);
    for (int i = 0; i < n; ++i) {
        cin >> points[i].x >> points[i].y;
    }
}

void printResult(const pair<Point, Point>& result) {
    cout << fixed << setprecision(4);
    cout << result.first.x << " " << result.first.y << " "
         << result.second.x << " " << result.second.y << endl;
}

int main() {
    vector<Point> points;
    while (true) {
        readPoints(points);
        if (points.empty()) break;

        pair<Point, Point> result = closestPair(points);
        printResult(result);
    }
    return 0;
}

pair<Point, Point> closestPair(vector<Point>& points) {
    if (points.size() == 1) {
        return {Point{INT_MIN, INT_MIN}, Point{INT_MAX, INT_MAX}};
    }
    if (points.size() == 2) {
        return {points[0], points[1]};
    }
    if (points.size() == 3) {
        return bruteForce(points);
    }

    sort(points.begin(), points.end(), [](const Point& a, const Point& b) {
        return a.x < b.x;
    });

    int mid = points.size() / 2;
    vector<Point> leftPoints(points.begin(), points.begin() + mid);
    vector<Point> rightPoints(points.begin() + mid, points.end());

    pair<Point, Point> pair1 = closestPair(leftPoints);
    pair<Point, Point> pair2 = closestPair(rightPoints);

    double dist1 = euclideanDistance(pair1.first, pair1.second);
    double dist2 = euclideanDistance(pair2.first, pair2.second);
    double minDist = min(dist1, dist2);

    vector<Point> strip;
    for (const auto& point : points) {
        if (abs(point.x - points[mid].x) < minDist) {
            strip.push_back(point);
        }
    }
    sort(strip.begin(), strip.end(), sortByY);

    pair<Point, Point> pair3 = closestPairInLine(strip, minDist);
    double dist3 = euclideanDistance(pair3.first, pair3.second);

    if (dist3 < minDist) {
        return pair3;
    } else if (dist1 < dist2) {
        return pair1;
    } else {
        return pair2;
    }
}

pair<Point, Point> closestPairInLine(vector<Point>& points, double distance) {
    pair<Point, Point> closestPair(Point{INT_MIN, INT_MIN}, Point{INT_MAX, INT_MAX});
    double minDist = distance;

    for (int i = 0; i < points.size(); ++i) {
        for (int j = i + 1; j < points.size() && points[j].y - points[i].y < minDist; ++j) {
            double dist = euclideanDistance(points[j], points[i]);
            if (dist < minDist) {
                closestPair = {points[i], points[j]};
                minDist = dist;
            }
        }
    }

    return closestPair;
}

double euclideanDistance(const Point& a, const Point& b) {
    double distX = a.x - b.x;
    double distY = a.y - b.y;
    return sqrt(distX * distX + distY * distY);
}

bool sortByY(const Point& a, const Point& b) {
    return a.y < b.y;
}

pair<Point, Point> bruteForce(vector<Point>& points) {
    pair<Point, Point> closestPair(Point{INT_MIN, INT_MIN}, Point{INT_MAX, INT_MAX});
    double minDist = euclideanDistance(closestPair.first, closestPair.second);

    for (int i = 0; i < points.size(); ++i) {
        for (int j = i + 1; j < points.size(); ++j) {
            double dist = euclideanDistance(points[i], points[j]);
            if (dist < minDist) {
                closestPair = {points[i], points[j]};
                minDist = dist;
            }
        }
    }

    return closestPair;
}
